sql/src/main/java/io/crate/analyze/where/WhereClauseNullReplacer.java

```diff
@@ -0,0 +1,111 @@
+import io.crate.execution.expression.operator.Operators;
+import io.crate.execution.expression.predicate.NotPredicate;
+import io.crate.metadata.FunctionInfo;
+
+/**
+ * Inside a where clause NULL values as logical operator arguments can be treated like boolean FALSE.
+ * As a result, the query can mostly be optimized and turned into a lucene query.
+ * Without this optimization chances are high that a genericFunctionFilter is used instead of a lucene query which
+ * results in poor performance (table scan -> filter).
+ * <p>
+ * <pre>
+ * Example:
+ *
+ *      NULL AND col1 = 1
+ *
+ *  can be handled as:
+ *
+ *      FALSE AND col1 = 1
+ *
+ *  which can be immediately normalized to:
+ *
+ *      FALSE -> NO-MATCH
+ * </pre>
+ *
+ * @implNote If a NOT predicate is encountered inside the tree, the current boolean to replace a NULL must be inverted
+ * for that leaf. Also traversing must be stopped if a conditional function is encountered as they will possible handle
+ * NULL values in a concrete way.
+ */
+public class WhereClauseNullReplacer {
+
+    private static final Visitor VISITOR = new Visitor();
+
+    public static Symbol replaceIfPossible(Symbol symbol) {
+        return VISITOR.process(symbol, new Context());
+    }
+
+    private static class Context {
+        boolean insideLogicalOperator = false;
+        boolean replaceNullBoolean = false;
+    }
+
+    private static class Visitor extends FunctionCopyVisitor<Context> {
+
+        @Override
+        public Symbol visitFunction(Function func, Context context) {
+            // we won't replace anything below a conditional function as NULL are possible handled
+            if (func.info().hasFeature(FunctionInfo.Feature.CONDITIONAL)) {
+                return func;
+            }
+            boolean replaceNullBoolean = context.replaceNullBoolean;
+            String functionName = func.info().ident().name();
+
+            // only operate inside logical operators
+            if (Operators.LOGICAL_OPERATORS.contains(functionName)) {
+                context.insideLogicalOperator = true;
+                if (NotPredicate.NAME.equals(functionName)) {
+                    // invert the boolean nulls are replaced with
+                    context.replaceNullBoolean = !replaceNullBoolean;
+                }
+            } else {
+                context.insideLogicalOperator = false;
+            }
+            Symbol newFunc = super.visitFunction(func, context);
+
+            // reset context
+            context.insideLogicalOperator = false;
+            context.replaceNullBoolean = replaceNullBoolean;
+
+            return newFunc;
+        }
+
+        @Override
+        public Symbol visitLiteral(Literal symbol, Context context) {
+            if (context.insideLogicalOperator && symbol.value() == null) {
+                return Literal.of(context.replaceNullBoolean);
+            }
+            return symbol;
+        }
+    }
+
+    private WhereClauseNullReplacer() {
```

Comment:
really minor: you can move the constructor as first method.

Reply:
I think this is fine as everything until the constructor is static. Please just have WhereClauseNullReplacer be "final" as it's a static construct.